torch/_numpy/_util.py

```diff
@@ -171,9 +171,17 @@ def typecast_tensors(tensors, target_dtype, casting):
     return tuple(typecast_tensor(t, target_dtype, casting) for t in tensors)
 
 
-def _try_convert_to_tensor(obj):
+def _try_convert_to_tensor(obj, dtype=None):
     try:
-        tensor = torch.as_tensor(obj)
+        try:
+            tensor = torch.as_tensor(obj, dtype=dtype)
+        except (TypeError, ValueError):
+            # is `obj` a list of tensors with nelem>1 ? E.g.:
+            # torch.as_tensor([torch.arange(3), torch.arange(3)])
+            # NB: as_tensor emits a ValueError for floating-point dtypes
+            #     and a TypeEror for integer dtypes
+            tensor = torch.stack(obj)
+            tensor = cast_if_needed(tensor, dtype)
```

Comment:
Rather than commenting why this works, let's make it explicit. Check if obj is an iterable and check whether all the objects in it are indeed tensors, then choose between torch.stack and torch.as_tensor depending on this.

Reply:
OK, a larger rework is on the way, https://github.com/pytorch/pytorch/pull/112521/files#diff-0daa64329e2d8648fc119f7809dc810c744508d345ea023116614ccc17e57dbeR1696